### PR TITLE
Feature/event system

### PR DESCRIPTION
**Pull request by HellPat**
Relates to https://github.com/phpab/phpab/pull/44

I built in a (very simple) event system.
There is only one Interface needed: `DispatcherInterface` and i have one implementation ready.

I's the most simple EventSystem i can think of.
No Priorization or anything else.

Test Coverage is still 100%. Travis is still failing on 5.5 (and i don't know why, hope it will get away once merged).

**Please Review**

**If everything is fine we could start with the following tasks**

---
### Implement Engine

how should we depend on the EventDispatcher. Maybe force a private method `EngineInterface`?
### Implement other Event Systems
- Symfony
- Zend
- Laravel
- ThePhpLeague

i would do this directly in in a `Bridge/` folder and add them to suggest.
This is not required for a first release.

### CODE REVIEW

**Comment by waltertamboer** [comment]
> Implement other Event Systems
> Symfony
> Zend
> Laravel
> ThePhpLeague
> i would do this directly in in a Bridge/ folder and add them to suggest.
> This is not required for a first release.

I would prefer to implement them in their own repositories.

File: src/Event/Dispatcher.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     *            The format looks like
+     *            [
+     *              'eventname' => [callable, callable, callable],
+     *              'eventname2' => [callable, callable, callable]
+     *            ]
+     */
+    private $listeners = [];
+
+    /**
+     * @param          $eventName The name of the event to listen for
```
Comment (waltertamboer) [inline src/Event/Dispatcher.php]: string

File: src/Event/ParticipationEvent.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace PhpAb\Event;
+
+use PhpAb\Test\TestInterface;
+use PhpAb\Variant\VariantInterface;
+
+/**
+ * Class ParticipationEvent
```
Comment (waltertamboer) [inline src/Event/ParticipationEvent.php]: This line you can remove, it's pretty obvious :)

File: src/Event/SubscriberInterface.php [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace PhpAb\Event;
+
+/**
+ * Interface SubscriberInterface
```
Comment (waltertamboer) [inline src/Event/SubscriberInterface.php]: Pretty obvious, you can remove this line.

**Comment by waltertamboer** [comment]
I'm done reviewing, I don't see any weird things. Looks good to me!

File: tests/Event/DispatcherTest.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        // Arrange
+        $dispatcher = new Dispatcher();
+
+        // Act
+        $result1 = $dispatcher->dispatch('event', null);
+        $result2 = $dispatcher->dispatch('event2', null);
+
+        // Assert
+        $this->assertNull($result1);
```
Comment (waltertamboer) [inline tests/Event/DispatcherTest.php:17]: Is this test needed? Dispatch always returns null and you;re checking that in every test again. Maybe you need one test: `testDispatchReturnsNull` and just check if it returns null. Than you don't have to add that assert in all the other tests.

File: src/Event/Dispatcher.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+     * @param callable $callable  The Callable to execute once the event takes place
+     */
+    public function addListener($eventName, callable $callable)
```
Comment (waltertamboer) [inline src/Event/Dispatcher.php:21]: We need a priority, else you cannot control if your listener should be executed before or after an event.

Comment (HellPat) [inline src/Event/Dispatcher.php:21]: Do we really need this for the simple implementation?
For other usecases we could just suggest symfony/zend one.

File: tests/Event/DispatcherTest.php (hunk 4 above)
Comment (HellPat) [inline tests/Event/DispatcherTest.php:17]: Done
